scripts/licenses/licenses/library.go

```diff
@@ -0,0 +1,200 @@
+// A library is a collection of one or more packages covered by the same license file.
+// Packages not covered by a license will be returned as individual libraries.
+// Packages covered by the same license as pkg will be ignored.
+// Standard library packages will be ignored.
+func Libraries(ctx *build.Context, pkg *build.Package) ([]*Library, error) {
+	pkgLicense, err := Find(pkg)
+	if err != nil {
+		return nil, err
+	}
+	deps := make(map[string]*build.Package)
+	if err := dependencies(ctx, pkg, deps); err != nil {
+		return nil, err
+	}
+	pkgsByLicense := make(map[string][]*build.Package)
+	for _, d := range deps {
+		if isStdLib(d) {
+			// No license requirements for the Go standard library.
+			continue
+		}
+		licensePath, err := Find(d)
+		if err != nil {
+			glog.Errorf("Failed to find license for %s: %v", d.ImportPath, err)
+		}
+		if licensePath == pkgLicense {
+			// Skip dependencies that share a license file with pkg.
+			// These are not external libraries.
+			continue
+		}
+		pkgsByLicense[licensePath] = append(pkgsByLicense[licensePath], d)
+	}
+	var libraries []*Library
+	for licensePath, pkgs := range pkgsByLicense {
+		if licensePath == "" {
+			// No license for these packages - return each one as a separate library.
+			for _, pkg := range pkgs {
+				libraries = append(libraries, &Library{
+					Packages: []*build.Package{pkg},
+				})
+			}
+			continue
+		}
+		libraries = append(libraries, &Library{
+			LicensePath: licensePath,
+			Packages:    pkgs,
+		})
+	}
+	return libraries, nil
+}
+
+// Name is the common prefix of the import paths for all of the packages in this library.
+func (l *Library) Name() string {
+	if len(l.Packages) == 0 {
+		return ""
+	}
+	if len(l.Packages) == 1 {
+		return l.Packages[0].ImportPath
+	}
+	var importPaths []string
+	for _, pkg := range l.Packages {
+		importPaths = append(importPaths, pkg.ImportPath)
+	}
+	sort.Strings(importPaths)
+	min, max := importPaths[0], importPaths[len(importPaths)-1]
+	lastSlashIndex := 0
+	for i := 0; i < len(min) && i < len(max); i++ {
+		if min[i] != max[i] {
+			return min[:lastSlashIndex]
+		}
+		if min[i] == '/' {
+			lastSlashIndex = i
+		}
+	}
+	return min
+}
+
+func (l *Library) String() string {
+	return l.Name()
+}
+
+// ClassifyLicense returns the name and type of license that applies to this library.
+func (l *Library) ClassifyLicense(classifier *Classifier) (string, Type, error) {
+	licenseName, err := classifier.Identify(l.LicensePath)
+	if err != nil {
+		return "", Unknown, fmt.Errorf("error identifying %q: %v", l.LicensePath, err)
+	}
+	licenseType := Unknown
+	switch licenseclassifier.LicenseType(licenseName) {
+	case "restricted":
+		licenseType = Restricted
+	case "reciprocal":
+		licenseType = Reciprocal
+	case "notice":
+		licenseType = Notice
+	case "permissive":
+		licenseType = Permissive
+	case "unencumbered":
+		licenseType = Unencumbered
+	case "FORBIDDEN":
+		licenseType = Forbidden
+	}
```

Comment:
Does the remapping of a string constants to enum values add enough value to make it worthwhile?

Reply:
It avoided me introducing bugs via typos when referencing license types, and makes the documentation for the license types easier to find. It's a fairly small amount of extra code, so I think it's worth it. I've done a bit of refactoring to reduce the amount of code (rather than mapping the strings to ints, I've just created const aliases for the known strings).

Reply:
Fair enough.